MAINT: Large overhead in some random functions

przemb opened the pull request:
Fixes https://github.com/numpy/numpy/issues/15460

- slow calls to np.dtype.name replaced with np.dtype,
- mtrand.pyx and _generator.pyx updated
- test `_test_warns_byteorder_` updated

eric-wieser commented on numpy/random/_generator.pyx:
```diff
@@ -454,20 +453,20 @@ cdef class Generator:
         else:
-            raise TypeError('Unsupported dtype "%s" for standard_exponential'
-                            % key)
+            raise TypeError('Unsupported dtype "%r" for standard_exponential'
```
I'd drop the quotes, they imply a string.
```suggestion
            raise TypeError('Unsupported dtype %r for standard_exponential'
```

przemb replied:
Thanks, updated.

eric-wieser commented on numpy/random/_generator.pyx:
```diff
@@ -559,39 +558,39 @@ cdef class Generator:
+                             'platform-independent byteorder, call byteswap '
+                             'when required.')
+        else:
+            raise TypeError('Non-native byteorder or unsupported dtype "%r" for randint' % _dtype)
```
If this line is reached, then you already know the byteorder is native, don't you?

przemb replied:
Thanks, I forgot to update the error msg.

charris commented:
Do these fixes change the byte stream?

seberg commented:
I am not sure this is a regression, since I think the legacy version did not have the dtype kwarg normally? But we might as well backport probably.

This should not change the byte streams, it should be purely maintenance in that sense. It might be nice to see if at least some of these are picked up by the benchmarks (if there are benchmarks drawing even a medium number from some of these distributions, I would expect so, however).

przemb committed:
MAINT: Large overhead in some random functions #15460

slow calls to np.dtype.name replaced with np.dtype,
mtrand.pyx and _generator.pyx updated,
test test_warns_byteorder updated

before:
%timeit rs.random(): 520 ns ± 33.1 ns per loop
%timeit rg.random(): 6.36 µs ± 222 ns per loop

after:
%timeit rs.random(): 453 ns ± 6.95 ns per loop
%timeit rg.random(): 594 ns ± 9.66 ns per loop

mattip commented on numpy/random/mtrand.pyx:
```diff
@@ -743,24 +732,31 @@ cdef class RandomState:
+            raise ValueError('Providing a dtype with a non-native byteorder '
+                             'is not supported. If you require '
+                             'platform-independent byteorder, call byteswap '
+                             'when required.')
```
This changes emitting a deprecation warning into an error, but according to our (still draft-status) [policy](https://numpy.org/neps/nep-0023-backwards-compatibility.html#policy) there should be at least a year between the start of the deprecation cycle until its removal..

przemb replied:
In case of _mtrand.pyx_ a warning was thrown, but in _generator.pyx ValueError was raised. I changed it to keep it consistent. I will roll back my modification.

eric-wieser commented on numpy/random/_generator.pyx:
```diff
@@ -559,39 +558,39 @@ cdef class Generator:
 
-        dt = np.dtype(dtype)
-        key = dt.name
-        if key not in _integers_types:
```
Is this variable used any more, or can it be removed?

przemb replied:
Do you mean __integers_types_? removed.

eric-wieser commented on numpy/random/mtrand.pyx:
```diff
@@ -724,17 +723,7 @@ cdef class RandomState:
-            raise TypeError('Unsupported dtype "%s" for randint' % key)
-        if not dt.isnative:
-            # numpy 1.17.0, 2019-05-28
-            warnings.warn('Providing a dtype with a non-native byteorder is '
```
I think you can keep this and add `dt = dt.newbyteorder()`

przemb replied:
Ok, good idea.

mattip commented:
I tried to set up my laptop for consistent benchmark running (configuring the system so I could do `sudo cpupower frequency-set --governor userspace`  then setting `sudo cpupower frequency-set -f 2.1GHz` which is much less than the top frequency), then running
```
python runtests.py --bench-compare HEAD
```
resulted in no significant changes.

eric-wieser commented:
Likely means we don't have useful benchmarks. Here's a micro benchmark whoing a big improvement:
```
In [10]: %timeit dt == dt2
84.2 ns ± 4.24 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [11]: %timeit dt == np.float32
191 ns ± 30.5 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [12]: %timeit dt.name == 'float32'
3.87 µs ± 688 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

eric-wieser commented on numpy/random/_generator.pyx:
```diff
@@ -559,39 +558,39 @@ cdef class Generator:
             ret = _rand_uint8(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
-        elif key == 'bool':
+        elif _dtype == np.bool_:
             ret = _rand_bool(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
```
A lookup table might be faster here, as:
```
_rand_int_funcs = {
    np.dtype(np.bool_): _rand_bool,
    ...
```
then
```
try:
    func = _rand_int_funcs[dt]
except KeyError:
    raise ValueError(...) from None
else:
    return func(...)

przemb replied:
I tried your suggestion, but I have an error during compilation. I don't have much experience with Cython, could you take a look? 
https://gist.github.com/przemb/95d6782dada6f1cfb05ff53c810577d4


btw: I thought about [overlading](https://www.python.org/dev/peps/pep-3124/#overloading-generic-functions), but I guess that it is not possible? (in order to keep backward compatibility)

rkern replied:
`_rand_bool` etc. are C functions (essentially) and cannot be used as dict values.

rkern replied:
The `if/elif` chain should be fine.

eric-wieser replied:
The advantage of the dict approach is that the `np.dtype` objects can be pre-computed, rather than `dt == np.float32` being expanded to `dt == np.dtype(np.float32)` on each call.

@rkern's explanation of why this doesn't work is correct, and it looks like it's more trouble than it's worth to make something like that work - so I agree that what you have is fine.

mattip commented:
Fixes #15460. I can see a big improvement on a microbenchmark:
```
$ python runtests.py --ipython
...
In [1]: rs = np.random.RandomState() 
   ...: rg = np.random.default_rng()
In [2]: %timeit rs.random()                                                                                                                                                              
191 ns ± 1.11 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [3]: %timeit rg.random()                                                                                                                                                              
2.54 µs ± 36.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
After this changeset (a15dc30)
```
In [2]: %timeit rs.random()                                                                                                                                                              
188 ns ± 2.07 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [3]: %timeit rg.random()                                                                                                                                                              
239 ns ± 0.968 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

przemb committed:
Changes suggested in review

eric-wieser commented on numpy/random/_bounded_integers.pyx.in:
```diff
@@ -51,16 +51,6 @@ cdef extern from "numpy/random/distributions.h":
 
 
-
-_integers_types = {'bool': (0, 2),
```
Did nothing use these values?

przemb replied:
I checked it, and I don't see any references. 
Please check: https://github.com/numpy/numpy/search?q=_integers_types&unscoped_q=_integers_types

eric-wieser commented on numpy/random/_generator.pyx:
```diff
@@ -312,13 +311,13 @@ cdef class Generator:
         cdef double temp
-        key = np.dtype(dtype).name
-        if key == 'float64':
+        _dtype = np.dtype(dtype)
```
Would normally spell this
```suggestion
        dtype = np.dtype(dtype)
```
Does cython not like that?

przemb replied:
If I change it, test are failing... seems that it is not proper.

eric-wieser replied:
Do you have a link to the CI failure?

mattip replied:
We need the original value so we can raise an error with it later. What about `descr` or `concrete_dtype`?

eric-wieser commented on numpy/random/_generator.pyx:
```diff
@@ -559,39 +558,39 @@ cdef class Generator:
+
 
         if size is None and dtype in (bool, int, np.compat.long):
             if np.array(ret).shape == ():
```
The line below this is is where the error is coming from. Perhaps for this function do
```
orig_dtype = dtype
```
and then here use` orig_dtype` instead of `dtype`

przemb replied:
did you mean: 
``
        dtype_ = None
        if size is None and dtype_ in (bool, int, np.compat.long):
            if np.array(ret).shape == ():
                return dtype_(ret)
        return ret
``

eric-wieser replied:
No, I meant:

* Change the functions that include these lines back to how they were in your previous commit, with the separate `dtype` and `dtype_` variables
* Leave the other functions with the better `dtype` name

przemb replied:
Thanks for the clarification.

eric-wieser commented on numpy/random/_generator.pyx:
```diff
@@ -559,39 +558,39 @@ cdef class Generator:
             high = low
```
Documentation at https://github.com/numpy/numpy/pull/15511/files#diff-f166b929b9f07d28bd1582546a65c87bR500 should be updated, perhaps something like:
```diff
- dtype : {str, dtype}, optional
+ dtype : dtype, optional
-           Desired dtype of the result. All dtypes are determined by their
-           name, i.e., 'int64', 'int', etc, so byteorder is not available
-           and a specific precision may have different C types depending
+           Desired dtype of the result. Byteorder must be native, and the result
+           dtype will match the size but perhaps not the underlying C type, depending
            on the platform. The default value is `np.int_`.
```

eric-wieser replied:
This comment still applies, there are multiple pieces of documentation like this

przemb replied:
You are right, I updated it. 
What about repeating part: "_All dtypes are determined by their name_". All occurrences of _dtype.name_ were replaced... it should be rather: "All dtypes are determined by their types", right?

eric-wieser replied:
I don't think there's any replacement for that sentence. I tried to capture the behavior with _"the result dtype will be equivalent but may not be identical"_. To elaborate on that, what I mean is that we guarantee
`func(dtype=dt).dtype == dt`, but not that `func(dtype=dt).dtype is dt`. I don't know if we want to mention that explicitly in the docs

Really, what we're trying to cover here is (on linux) someone passing `np.longlong`, and being surprised when they get back `np.int_`, even though both are 64-bit.

Maybe that's not interesting and we should just remove the remark entirely. Perhaps another maintainer can comment on whether we need it.

mattip replied:
it is a bit confusing. I think we can remove ", and the result dtype will be equivalent but may not be identical", no need to dive into the weeds here.

przemb replied:
Ok, updated.

przemb committed:
Documentation updated.

eric-wieser reviewed (changes requested):
Docs still need updating as mentioned in https://github.com/numpy/numpy/pull/15511/files#r375407384

eric-wieser commented on numpy/random/_generator.pyx:
```diff
@@ -278,10 +277,10 @@ cdef class Generator:
             single value is returned.
-        dtype : {str, dtype}, optional
+        dtype : dtype, optional
             Desired dtype of the result, either 'd' (or 'float64') or 'f'
```
```suggestion
            Desired dtype of the result, either ``np.float64`` or ``np.float32``.
```

eric-wieser replied:
Same throughout.

eric-wieser commented on numpy/random/_generator.pyx:
```diff
@@ -278,10 +277,10 @@ cdef class Generator:
             Desired dtype of the result, either 'd' (or 'float64') or 'f'
-            (or 'float32'). All dtypes are determined by their name. The
-            default value is 'd'.
+            (or 'float32'). Byteorder must be native.
```
```suggestion
            Byteorder must be native.
```

eric-wieser commented on numpy/random/mtrand.pyx:
```diff
@@ -643,7 +642,7 @@ cdef class RandomState:
     def randint(self, low, high=None, size=None, dtype=int):
         """
-        randint(low, high=None, size=None, dtype='l')
+        randint(low, high=None, size=None, dtype=int)
```
```suggestion
        randint(low, high=None, size=None, dtype=np.int_)
```
As written this is a behavior change

eric-wieser replied:
Oh wait, this was actually correct before. Sorry, I confused the docstring with the function def.

eric-wieser commented on numpy/random/mtrand.pyx:
```diff
@@ -670,10 +669,8 @@ cdef class RandomState:
-            and a specific precision may have different C types depending
-            on the platform. The default value is `np.int_`.
+            Desired dtype of the result. Byteorder must be native.
+            The default value is int.
```
```suggestion
            The default value is ``np.int_``.
```

eric-wieser reviewed (approved):
Thanks for working through the doc fixes!

mattip commented on numpy/random/_generator.pyx:
```diff
@@ -496,11 +493,9 @@ cdef class Generator:
-            on the platform. The default value is `np.int_`.
+        dtype : dtype, optional
+            Desired dtype of the result. Byteorder must be native.
+            The default value is `np.int_`.
```
The default (line 469) is np.int64 but default is documented as np.int_? That seems wrong.

mattip replied:
Are there more mistakes like this across the docstrings?

przemb replied:
updated

eric-wieser commented on numpy/random/mtrand.pyx:
```diff
@@ -643,7 +642,7 @@ cdef class RandomState:
     def randint(self, low, high=None, size=None, dtype=int):
         """
-        randint(low, high=None, size=None, dtype='l')
+        randint(low, high=None, size=None, dtype=np.int_)
```
Apologies for the incorrect advice earlier - this should have been `int` after all, as should 673. These repeated signature lines are confusing.

przemb replied:
Thanks :D

eric-wieser reviewed (approved):
LGTM

mattip commented:
Thanks @przemb
